Zoom To now acts like Find Next

nonbinarysolo opened the pull request:
The Zoom To dialog box has been replaced and now acts as a sort of Find Next feature. This enables the user to search for complete object labels or substrings found in object labels and then scroll through the results (it also provides utilities for skipping hidden objects and wraparound search). It also binds CTRL+G as a "find next" shortcut in the main window so that matches can be inspected without the need to refer back to the dialog box. These changes resolve Issue #36.

jefft0 commented on zoomto-window.cpp:
```diff
@@ -0,0 +1,290 @@
+//_/_/ Copyright (c) 2018-2022 Jeff Thompson
+//_/_/ Copyright (c) 2018-2022 Kristinn R. Thorisson
+//_/_/ Copyright (c) 2018-2022 Icelandic Institute for Intelligent Machines
+//_/_/ Copyright (c) 2021 Karl Asgeir Geirsson
```
This file is original by you. Maybe this was copy-pasted from another file. If you wish, you can replace Karl with
Copyright (c) 2023 .... and use your name as you want it to appear. Also please change the "general" copyright assignees years to 2018-2023.

nonbinarysolo replied:
Good point! I just dropped that in when I first created that file, I totally forgot to update it. I just pushed a new commit that should fix this

jefft0 commented on zoomto-window.hpp:
```diff
@@ -0,0 +1,128 @@
+//_/_/ Copyright (c) 2018-2022 Jeff Thompson
+//_/_/ Copyright (c) 2018-2022 Kristinn R. Thorisson
+//_/_/ Copyright (c) 2018-2022 Icelandic Institute for Intelligent Machines
+//_/_/ Copyright (c) 2021 Karl Asgeir Geirsson
```
Please see the comment for zoomto-window.cpp.

jefft0 commented on zoomto-window.cpp:
```diff
@@ -0,0 +1,290 @@
+
+namespace aera_visualizer {
+
+  ZoomToWindow::ZoomToWindow(AeraVisualizerWindow *parent, ReplicodeObjects& replicodeObjects) {
```
If you add `: QDialog(parent)` then the dialog will not have a separate entry on the Windows task bar (which is what we want, I think.) See: https://stackoverflow.com/a/40240764

nonbinarysolo replied:
Okay, sounds reasonable. This changes the behavior a little bit since the Zoom To window now stays on top while it's open, but I think that makes sense for this kind of Find feature. You can still use the main window as normal, it'll just sit there and give an overview of the search progress I suppose.

nonbinarysolo commented:
> Nice flashing items!

Thanks!

> A possible bug:
> 
>     * In AERA, run the example hand-grab-sphere.replicode
> 
>     * Load it in AERA Visualizer. Click the "step" button to show the first frame.
> 
>     * The first frame has four items. One of the items is fact_123. Click the find button
> 
>     * In the find dialog, type `123` and click Find Next .
> 
> 
> Expected: The find dialog should say "1 out of 1 matches". Actual: The find dialog says "2 out of 8 matches".

I'll take a look into that bug; looks like part of it has to do with the find matches ignoring the current time setting, but it may also be detecting some background items that are never visible. There should be a way around this, it may just take a little while.

nonbinarysolo commented:
I just pushed a new commit that changes the way matches are stored and presented. The `findNext` function now sorts out any items without a valid AeraGraphicsItem which prevents background items or items that have not occurred yet from showing up as a match. This list is refreshed any time the search term is updated or the current position in time is changed.

jefft0 commented:
Feature request: 

* Rename "Zoom To" box to "Find" and add a "Zoom To" checkbox.
* Find backwards button as well as forwards.

nonbinarysolo commented:
I just made some changes that let you search backwards and choose whether or not to zoom to the found item. I also fixed up the sorting helper (matches are ordered by time and then OID), adjusted a few shortcuts, and removed the somewhat redundant "Close" button.

jefft0 commented:
We're getting closer. On my computer there are still a couple bugs. See this video:
https://www.loom.com/share/6a3c9d59b19f403f987bf6d3c4283b50

nonbinarysolo commented:
I figured out the bug, looks like there was a weird edge case that allowed it to "skip" hidden items before revealing them all at once. I've updated some of the behaviors so that shouldn't be possible anymore. I also rebased the branch to catch up with some changes on dev.

jefft0 commented:
This is much better. I'm still getting occasional crashes in weird edge cases, like if I'm on an "essence" item in the find results, then I unclick "Essence Facts" to hide it, then click Find Next. If you want to dive in, then I can make a video to show the case (or you can try to reproduce). Otherwise, we can merge the code which generally works great and open an other issue for the bug. It's up to you.

nonbinarysolo commented:
Strange, I haven't been able to reproduce that. I'd be down to merge and investigate it separately, though, since it might involve adding in event handlers for the checkboxes which may change the code quite a bit.

nonbinarysolo commented:
Sounds good!

jefft0 commented:
Please squash the commits before the meeting, so that we are reviewing the code that will be merged. :-)
